### PR TITLE
hsm: update libp11 patch

### PR DESCRIPTION
**Pull request by istepic**
Some PKCS#11 implementations don't support all the features libp11 is using. Some of them are removed.

### CODE REVIEW

File: dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch [hunk 1]
```diff
@@ -368,7 +369,32 @@ index ec7f279..57cba43 100644
  
  	/* pubkey attributes */
  	pkcs11_addattr(&pubtmpl, CKA_ID, id, id_len);
-@@ -310,6 +320,105 @@ int pkcs11_generate_key(PKCS11_SLOT_private *slot, int algorithm, unsigned int b
+ 	if (label)
+ 		pkcs11_addattr_s(&pubtmpl, CKA_LABEL, label);
+ 	pkcs11_addattr_bool(&pubtmpl, CKA_TOKEN, TRUE);
+-	pkcs11_addattr_bool(&pubtmpl, CKA_ENCRYPT, TRUE);
```
Comment (garw) [inline dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch:381]: This is for RSA keys? 
I'm not sure if RSA keys can do derive. you can usually use ENCRYPT

File: dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch [hunk 2]
```diff
@@ -368,7 +369,32 @@ index ec7f279..57cba43 100644
...
++	pkcs11_addattr_bool(&pubtmpl, CKA_DERIVE, TRUE);
+ 	pkcs11_addattr_bool(&pubtmpl, CKA_VERIFY, TRUE);
+-	pkcs11_addattr_bool(&pubtmpl, CKA_WRAP, TRUE);
+ 	pkcs11_addattr_var(&pubtmpl, CKA_MODULUS_BITS, num_bits);
+ 	pkcs11_addattr(&pubtmpl, CKA_PUBLIC_EXPONENT, public_exponent, 3);
```
Comment (garw) [inline dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch]: public exponent 3 is strongly discouraged. Why do we set this? Do we really need to? The common public exponent for RSA is nowadays 0x10001

File: dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch [hunk 3]
```diff
@@ -368,7 +369,32 @@ index ec7f279..57cba43 100644
...
+ 
+@@ -286,12 +295,11 @@ int pkcs11_generate_key(PKCS11_SLOT_private *slot, int algorithm, unsigned int b
+ 	pkcs11_addattr(&privtmpl, CKA_ID, id, id_len);
+ 	if (label)
+ 		pkcs11_addattr_s(&privtmpl, CKA_LABEL, label);
+-	pkcs11_addattr_bool(&privtmpl, CKA_TOKEN, TRUE);
+ 	pkcs11_addattr_bool(&privtmpl, CKA_PRIVATE, TRUE);
++	pkcs11_addattr_bool(&privtmpl, CKA_TOKEN, TRUE);
+ 	pkcs11_addattr_bool(&privtmpl, CKA_SENSITIVE, TRUE);
+-	pkcs11_addattr_bool(&privtmpl, CKA_DECRYPT, TRUE);
```
Comment (garw) [inline dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch]: s.a. This probably needs to depend on the algorithm

File: dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch [hunk 4]
```diff
@@ -417,36 +443,31 @@ index ec7f279..57cba43 100644
...
-+	ecdsa_params_len = i2d_ASN1_OBJECT(curve_obj, NULL);
-+	ecdsa_params = (unsigned char *)OPENSSL_malloc(ecdsa_params_len);
-+	if (!ecdsa_params)
++	ec_params_len = i2d_ASN1_OBJECT(curve_obj, NULL);
++	ec_params = (unsigned char *)OPENSSL_malloc(ec_params_len);
++	if (!ec_params)
 +		return -1;
-+	tmp = ecdsa_params;
++	tmp = ec_params;
 +	i2d_ASN1_OBJECT(curve_obj, &tmp);
```
Comment (garw) [inline dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch]: readability: Why do we pass tmp here and not ec_params? I just looked up what I believe to be the reason for this in OpenSSLs documentation. tmp is a bad variable name if the purpose isn't obvious and it deserves a comment here to explain whats going on with strange OpenSSL API.

return value check?

File: dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch [hunk 5]
```diff
@@ -417,36 +443,31 @@ index ec7f279..57cba43 100644
...
 +
 +	/* pubkey attributes */
 +	pkcs11_addattr(&pubtmpl, CKA_ID, id, id_len);
 +	if (label)
 +		pkcs11_addattr_s(&pubtmpl, CKA_LABEL, label);
 +	pkcs11_addattr_bool(&pubtmpl, CKA_TOKEN, TRUE);
-+	pkcs11_addattr_bool(&pubtmpl, CKA_DERIVE, FALSE);
-+	pkcs11_addattr_bool(&pubtmpl, CKA_WRAP, FALSE);
++	pkcs11_addattr_bool(&pubtmpl, CKA_DERIVE, TRUE);
```
Comment (garw) [inline dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch]: I don't think we need DERIVE on the pubkey. Not sure if this even works.

File: dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch [hunk 6]
```diff
@@ -474,6 +495,19 @@ index ec7f279..57cba43 100644
...
+diff --git a/src/p11_load.c b/src/p11_load.c
+index e89b0c7..3f69cec 100644
+--- a/src/p11_load.c
++++ b/src/p11_load.c
+@@ -81,7 +81,7 @@ int pkcs11_CTX_load(PKCS11_CTX *ctx, const char *name)
+ 	/* Tell the PKCS11 to initialize itself */
+ 	memset(&args, 0, sizeof(args));
+ 	/* Unconditionally say using OS locking primitives is OK */
+-	args.flags |= CKF_OS_LOCKING_OK;
++//	args.flags |= CKF_OS_LOCKING_OK;
```
Comment (garw) [inline dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch]: If we comment this out we can remove it entirely?

**Review by garw — CHANGES_REQUESTED** [review]

File: dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch (hunk 1 above)
Comment (istepic) [reply dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch:381]: Removed derive

File: dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch (hunk 2 above)
Comment (istepic) [reply dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch]: 3 is the length. Changed it to 0x10001

File: dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch (hunk 4 above)
Comment (istepic) [reply dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch]: Documented and return value checked

File: dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch (hunk 5 above)
Comment (istepic) [reply dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch]: Removed

File: dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch (hunk 6 above)
Comment (istepic) [reply dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch]: yes

File: dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch [hunk 7]
```diff
@@ -417,36 +448,38 @@ index ec7f279..57cba43 100644
...
 +	/* privkey attributes */
 +	pkcs11_addattr(&privtmpl, CKA_ID, id, id_len);
 +	if (label)
 +		pkcs11_addattr_s(&privtmpl, CKA_LABEL, label);
-+	pkcs11_addattr_bool(&privtmpl, CKA_TOKEN, TRUE);
 +	pkcs11_addattr_bool(&privtmpl, CKA_PRIVATE, TRUE);
++	pkcs11_addattr_bool(&privtmpl, CKA_TOKEN, TRUE);
 +	pkcs11_addattr_bool(&privtmpl, CKA_SENSITIVE, TRUE);
 +	pkcs11_addattr_bool(&privtmpl, CKA_DERIVE, TRUE);
-+	pkcs11_addattr_bool(&privtmpl, CKA_UNWRAP, FALSE);
```
Comment (johnfxgalea-e) [inline dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch:447]: Overall, can there be a comment explaining rationale for disabling/enabling certain attributes? It is hard to follow what your intent is, backed by the standard.

Keep in mind that some PKCS#11 attributes have default values so it could be that removal does not imply falsifying the attribute.

For instance, we don't want wrapping functionality, right? so leaving it false explicitly is a good thing no?

Comment (istepic) [reply dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch:447]: Where do we stop in that case? There are many object attributes. Do we want to have an explicit opinion on each and every one or do we leave some to the default values of PKCS#11 implementation? Also, problems could arise from PKCS#11 implementation not supporting setting a particular attribute. I think an argument can be made to make these flags configurable on the API level but it's too late for that now and we can think about that in the future when we move to OpenSSL 3.

I will update the commit message trying to explain the change.

**Review by garw — APPROVED** [review]

**Review by moh299814 — APPROVED** [review]
Based on http://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/errata01/os/pkcs11-curr-v2.40-errata01-os-complete.html#_Toc441850449 and related templates the flags seems fine, but it should be good to have some comments for the flags or at least a link to the related section in the reference document.

**Comment by istepic** [comment]
> Based on http://docs.oasis-open.org/pkcs11/pkcs11-curr/v2.40/errata01/os/pkcs11-curr-v2.40-errata01-os-complete.html#_Toc441850449 and related templates the flags seems fine, but it should be good to have some comments for the flags or at least a link to the related section in the reference document.

I added a comment about what the flags are used for. People using libp11 should be able to find the rest in the standard.

File: dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch (hunk 7 above)
Comment (johnfxgalea-e) [reply dockerfiles/feature-support/hsm-patches/0001-Introduce-generic-keypair-generation-interface-and-e.patch:447]: Commit message makes things clearer, thanks.

**Review by johnfxgalea-e — APPROVED** [review]
